servicetalk-grpc-internal/src/main/java/io/servicetalk/grpc/internal/StatusMessageUtils.java

```diff
@@ -0,0 +1,152 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Copyright 2024 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.internal;
+
+import io.servicetalk.http.api.HttpHeaders;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.buffer.api.CharSequences.newAsciiString;
+
+/**
+ * Provides utilities around percent-encoding and decoding the GRPC status message.
+ * <p>
+ * Note that much of the actual encoding and decoding logic is borrowed from the {@code io.grpc.Status} class,
+ * specifically the {@code io.grpc.Status#StatusMessageMarshaller}.
+ */
+public final class StatusMessageUtils {
+
+    public static final CharSequence GRPC_STATUS_MESSAGE = newAsciiString("grpc-message");
+
+    private static final byte[] HEX =
+            {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+    private StatusMessageUtils() {
+        // singleton
+    }
+
+    /**
+     * Sets and potentially encodes the given status message.
+     *
+     * @param headers the headers on where to set the message.
+     * @param message the message to set.
+     */
+    public static void setStatusMessage(final HttpHeaders headers, final CharSequence message) {
+        final boolean needsEncoding = message.chars().anyMatch(c -> isEscapingChar((char) c));
+        headers.set(GRPC_STATUS_MESSAGE, needsEncoding ? encodeMessage(message) : message);
+    }
+
+    /**
+     * Tries to read the status message from the {@link HttpHeaders} and percent-decode if necessary.
+     *
+     * @param headers the headers to load and decode the status message from.
+     * @return the decoded status message, or null of no message found in the header provided.
+     */
+    @Nullable
+    public static CharSequence getStatusMessage(final HttpHeaders headers) {
+        final CharSequence msg = headers.get(GRPC_STATUS_MESSAGE);
+        if (msg == null) {
+            return null;
+        }
+
+        final int msgLength = msg.length();
+        for (int i = 0; i < msgLength; i++) {
+            char b = msg.charAt(i);
+            if (b < ' ' || b >= '~' || (b == '%' && i + 2 < msgLength)) {
+                // fast-pathing did not work, perform actual decoding
+                return decodeMessage(msg);
+            }
```

Comment:
Good point - the current behavior is what io.grpc is doing. I think we should align with their behavior so we are not causing surprise with our users?

for what it's worth, I think your approach would be technically valid, since the spec is like "if you see something odd on decoding, you can do whatever as long as you don't discard":

> The value portion of Status-Message is conceptually a Unicode string description of the error, physically encoded as UTF-8 followed by percent-encoding. Percent-encoding is specified in [RFC 3986 §2.1](https://tools.ietf.org/html/rfc3986#section-2.1), although the form used here has different restricted characters. When decoding invalid values, implementations MUST NOT error or throw away the message. At worst, the implementation can abort decoding the status message altogether such that the user would received the raw percent-encoded form. Alternatively, the implementation can decode valid portions while leaving broken %-encodings as-is or replacing them with a replacement character (e.g., '?' or the Unicode replacement character).